### PR TITLE
Implement SchemaMigrator.

### PR DESCRIPTION
**Pull request by swankjesse**
I'm deliberately avoiding adding a more capable dependency like Flyway. This
should be upwards-compatible if ever we need to support multiple databases.

### CODE REVIEW

File: misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import javax.persistence.PersistenceException
+
+private val logger = getLogger<SchemaMigrator>()
+
+/**
+ * Manages **available** and **applied** schema migrations.
+ *
+ * Available schema migrations are SQL files in the datasource's `migrations_path` directory. Each
+ * file should contain SQL statements terminated by a `;`. The files should be named like
+ * `v100__exemplar.sql` with a `v`, an integer version, two underscores, a description, and the
```
Comment (shellderp) [inline misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt]: is this a tradition we want to carry forward?

File: misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  fun availableMigrations(): NavigableMap<Int, String> {
+    val resources = ResourceLoader.list(config.migrations_path)
+    return TreeMap(resources.associateBy { resourceVersion(it) })
+  }
+
+  /** Creates the `schema_version` table if it does not exist. Returns the applied migrations. */
+  fun initialize(): NavigableSet<Int> {
+    try {
+      val result = appliedMigrations()
+      logger.info {
```
Comment (shellderp) [inline misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt]: neat

File: misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+  /**
+   * Returns the versions of applied migrations. Throws a [PersistenceException] if the migrations
+   * table has not been initialized.
+   */
+  fun appliedMigrations(): NavigableSet<Int> {
+    sessionFactory.openSession().use { session ->
+      @Suppress("UNCHECKED_CAST") // createNativeQuery returns a raw Query.
+      val query = session.createNativeQuery("SELECT version FROM schema_version") as Query<String>
+      return TreeSet(query.list().map { it.toInt() })
```
Comment (shellderp) [inline misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt]: here you create a List then a TreeSet, but you could mapTo the TreeSet directly
```
return query.list().mapTo(TreeSet()) { it.toInt() }
```
I'm not picky though

File: misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt (hunk 1 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt]: I think it’s good enough!

File: misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt (hunk 3 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt]: I think it does the same thing underneath?

File: misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt (hunk 1 above)
Comment (eeSeeGee) [reply misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt]: Any reason why version rather than timestamp?

Comment (tirsen) [reply misk/src/main/kotlin/misk/hibernate/SchemaMigrator.kt]: Timestamp in milliseconds would preserve some form of order while still providing for super low probability of conflicts. The only drawback I can see is that the file name becomes rather long!

**Pull request merged**